difference in `-CC` behavior between aro and clang

This bit of `preserve comments in macros.c`:

https://github.com/Vexu/arocc/blob/8f4df188ac8256fc4cd9bc069a3b8dd68fd62b68/test/cases/preserve%20comments%20in%20macros.c#L13-L15

causes an error in Aro but not in clang

with `clang -E -CC`:

```
hello 1
```

with `arocc -E -CC`:

```
test.c:1:14: error: pasting formed '/*foo*/1', an invalid preprocessing token
#define e(p) p ## 1
             ^
test.c:2:8: note: expanded from here
e(hello/*foo*/)
       ^
```

---

Partially related to https://github.com/Vexu/arocc/issues/1060 / https://github.com/Vexu/arocc/pull/1061 in that there are other cases that differ from clang but will be fixed by https://github.com/Vexu/arocc/pull/1061

For example:

```c
#if /*foo*/1
foo
#endif
```

with clang (and aro with https://github.com/Vexu/arocc/pull/1061 applied):

```
foo
```

with master branch aro:

```
test.c:1:5: error: invalid token at start of a preprocessor expression
#if /*foo*/1
    ^
```

--- a/src/aro/Preprocessor.zig
+++ b/src/aro/Preprocessor.zig
@@ -2794,6 +2794,20 @@ fn pasteTokens(pp: *Preprocessor, lhs_toks: *ExpandBuf, rhs_toks: []const TokenW
         return bufCopyTokens(gpa, lhs_toks, rhs_toks, &.{});
     };
 
+    // preserved comments (via -CC) inhibit pasting
+    if (lhs.id == .comment and lhs.flags.is_macro_arg) {
+        if (lhs_toks.items.len != 0 and rhs_toks.len != 0) {
+            switch (lhs_toks.items[lhs_toks.items.len - 1].id) {
+                .macro_ws, .comment => {},
+                else => switch (rhs_toks[0].id) {
+                    .macro_ws, .comment => {},
+                    else => try lhs_toks.append(gpa, .{ .id = .macro_ws, .loc = lhs.loc }),
+                },
+            }
+        }
+        return bufCopyTokens(gpa, lhs_toks, rhs_toks, &.{});
+    }
+
     var rhs_rest: u32 = 1;
     const rhs = for (rhs_toks) |rhs| {
         if ((pp.comp.langopts.preserve_comments_in_macros and rhs.id == .comment) or

--- a/test/cases/expanded/preserve comments paste whitespace.c
+++ b/test/cases/expanded/preserve comments paste whitespace.c
@@ -1,0 +1,2 @@
+hello world
+hello 1

--- a/test/cases/preserve comments in macros.c
+++ b/test/cases/preserve comments in macros.c
@@ -35,6 +35,4 @@ args = -CC -fms-extensions
 
 preserve comments in macros.c:1:15: error: pasting formed '/*foo*//*bar*/', an invalid preprocessing token
 preserve comments in macros.c:4:13: error: pasting formed '/*foo*//*bar*/', an invalid preprocessing token
-preserve comments in macros.c:13:14: error: pasting formed '/*foo*/1', an invalid preprocessing token
-preserve comments in macros.c:15:8: note: expanded from here
 */

--- a/test/cases/preserve comments paste whitespace.c
+++ b/test/cases/preserve comments paste whitespace.c
@@ -1,0 +1,12 @@
+#define paste(p, q) p ## q
+#define suffix(p) p ## 1
+
+paste(hello/*trailing*/, world)
+suffix(hello/*trailing*/)
+
+#if 0 // prevent this from showing up in expanded output
+/** manifest:
+expand
+args = -CC
+*/
+#endif

